Rubi progress and tests

ashishkg0022 opened the pull request:
<!-- BEGIN RELEASE NOTES -->
* integrals
    * Made rubi stable.
    * Added missing utility functions
    * Added large test set

<!-- END RELEASE NOTES -->

asmeurer committed:
Make lambdify decorator a little faster

I pulled out the common functionality to be evaluated once, and used array()
instead of asarray() (they are equivalent, but array() is faster).

This is a partial fix to #14671. It isn't as fast as it used to be yet. I
don't know how to make the decorator faster, other than code generating the
array wrapping directly, which would require improvements to lambdify to work
(lambdify would need to be able to code generation full Python functions).

ashishkg0022 committed:
fixed_order

ashishkg0022 committed:
Merge branch 'separate_cons' into rubi_progress

ashishkg0022 committed:
fixed minors

ashishkg0022 committed:
fixed minor issues to pass test_1_2

ashishkg0022 committed:
updated utility functions

ashishkg0022 committed:
improved defn of AppellF1

ashishkg0022 committed:
corrected positiveQ

ashishkg0022 committed:
Updated for keeping track of rules applied

ashishkg0022 committed:
corrected trinomialMatchQ

ashishkg0022 committed:
corrected binomialMatchQ

ashishkg0022 committed:
updated all

ashishkg0022 commented:
In original test suite for trinomial products there are 5 files. In sympy format I have kept it here. https://github.com/ashishkg0022/rubi-structure/tree/master/test/trinomial . This contains all trinomial tests from the rubi test-suite.
I have added `test_trinomial_passed`  in this PR . This contains tests which are passed for trinomial_1 . Almost 600 test have passed for trinomial_1 . I started testing trinomial_2 . Then I encountered with issues like 

- `With` Function is not handled properly. It should be parsed correctly.

- Some times `Matchpy` wrongly matches `x` with some integer

- and some small issues.
The most important for now is fixing the `With` issue. Then majority of trinomial_2 should pass.

ashishkg0022 commented:
ping @Upabjojr  @parsoyaarihant

ashishkg0022 commented:
Also  
```python
In [1]: from sympy.abc import a,b,c,x, e, d

In [2]: b*e - c*d
Out[2]: b*e - c*d
```
but in mathematica 
```
In[1]:= b*e - c*d
Out[1]= -cd+be
 ```
This results in difference because there is a function `PosQ` which treats `b*e -c*d` True and `-c*d +b*e` False

ashishkg0022 committed:
indentation corrected

ashishkg0022 committed:
minor correct

parsoyaarihant commented:
> Some times Matchpy wrongly matches x with some integer
This should be reported to MatchPy developers. Can you share a sample case where this fails?

parsoyaarihant commented:
Regadring difference in outout of SymPy and Mathematica, Mathematica sorts the expression canonically. See: http://reference.wolfram.com/language/ref/Orderless.html http://reference.wolfram.com/language/ref/Sort.html 

You can use `Sort` in the utility function to address this issue.

ashishkg0022 commented:
> You can use Sort in the utility function to address this issue.

It handles rest of the cases . But specifically the case `b*e -c*d` is not sorted by the `Sort` function too.

ashishkg0022 committed:
corrected PolyQ

ashishkg0022 committed:
corrected With definitions

ashishkg0022 committed:
minor

ashishkg0022 committed:
Solved exceptions and wrong output

ashishkg0022 committed:
updated utility functions

ashishkg0022 committed:
added test for trinomial

ashishkg0022 committed:
updated cons

ashishkg0022 committed:
major Changes in utility function

ashishkg0022 committed:
added tests for miscellaneous algebra

ashishkg0022 committed:
added rules and tests for exponentials

ashishkg0022 committed:
solved merge conflict

ashishkg0022 committed:
renamed trinomials

ashishkg0022 committed:
added test in submodule

ashishkg0022 committed:
blacklisted rubi_tests

ashishkg0022 committed:
fixed few parsing issues

ashishkg0022 committed:
Major changes in test_parse and test_logarithms

ashishkg0022 committed:
updated polynomial qouotient

ashishkg0022 committed:
improved for trigo rules

ashishkg0022 committed:
merged master

ashishkg0022 committed:
added small test cases for each type of rule

ashishkg0022 committed:
added old tests in rubi_tests

ashishkg0022 committed:
updated Part

Upabjojr commented:
would you feel like separating the parsetools and the generated code in two separate PRs?

ashishkg0022 commented:
Closing this Pr and opening two new PRs. One including generated rules, constraints, tests . Other including parsetools and utility_functions ?

Upabjojr commented:
Sorry, ignore my previous message. It's better to stay here.

Upabjojr commented:
Some files are automatically generated, right?

Upabjojr commented:
In the file ` sympy/integrals/rubi/parsetools/header.py.txt`, better put a warning that the code to automatically generated, and not to edit.

Upabjojr commented:
Are you using my repository to generate the code (with the code I have parsed from Mathematica)? Or did you use your own?

ashishkg0022 commented:
Yes

ashishkg0022 commented:
> Are you using my repository to generate the code (with the code I have parsed from Mathematica)? Or did you use your own?

I have just updated last year's parsetools.

Upabjojr commented:
The file `/usr/people/home/bonazzi/sympy/sympy/integrals/rubi/__init__.py` needs more details.

ashishkg0022 commented:
Your repository contains all parsed rules in a single file. I have parsed them separately. Should I create a repo for that ?

Upabjojr commented:
> Your repository contains all parsed rules in a single file. I have parsed them separately. Should I create a repo for that ?

We need a way to allow a user to simply launch a command and generate the rules. Like automatic retrieval.

All instructions should be written in `/usr/people/home/bonazzi/sympy/sympy/integrals/rubi/__init__.py`

Upabjojr commented:
Even the Mathematica commands should be described.

ashishkg0022 commented:
> We need a way to allow a user to simply launch a command and generate the rules. Like automatic retrieval.

Getting the `DownValues` has to be done in steps. I loaded all rules (files of each type) one by one then generated the `DownValues` for each. After we get the downvalues, everything can be generated by one command.

Upabjojr commented:
can you put that into a script?

ashishkg0022 commented:
script for generating the rule files from downvalues?

Upabjojr commented:
Yes, also a script for generating the DownValues... let's create a nice documentation with all steps to parse the rules.

ashishkg0022 commented:
Should I create something like `rubi_parsing_guide.md` . 
Also will we be including those script files in sympy or instead of files, I write the code in the `rubi_parsing_guide.md` ?

ashishkg0022 committed:
modified parsetools and added guide.md

ashishkg0022 commented:
I have added this https://github.com/ashishkg0022/sympy/blob/5dacc8d6e6df150e44a3d025c42294c0015bc3e1/sympy/integrals/rubi/parsetools/rubi_parsing_guide.md
along with `generate_rules.py` . 

I think it will be good If I make a function which parses the test cases too from mathematica format. We should not rely upon maple format. Should I add it in `rubi_test` submodule?

Upabjojr commented:
> We should not rely upon maple format. Should I add it in rubi_test submodule?

Yes, add everything that may be useful.

ashishkg0022 committed:
added parser for test

ashishkg0022 commented:
I added `generate_tests.py` for parsing tests and tested test cases of `error_functions` with this parser. It is working fine. I added the steps in `rubi_parsing_guide.md`. Please check this
https://github.com/ashishkg0022/sympy/blob/97db9b310dc90900f6ff005a82c791002711493f/sympy/integrals/rubi/parsetools/rubi_parsing_guide.md

ashishkg0022 committed:
minor typo

ashishkg0022 committed:
updated for travis

ashishkg0022 committed:
Added missing tests in utility function

Abdullahjavednesar commented:
@ashishkg0022 have you implemented all the utility functions? What are you working on next? When do you plan to start with the documentation?

ashishkg0022 commented:
Yes, I have implemented all. There is a missing implementation in `apart`. I have asked on gitter about how should I proceed next

ashishkg0022 committed:
added some comments

Abdullahjavednesar commented:
That's good, so what do you plan to do next?

ashishkg0022 committed:
added examples and docstrings of some important function

ashishkg0022 commented:
I will add more instructions and documentation in this PR. I want this to merged as soon as possible so that I can move on to update utility functions of the latest version.

ashishkg0022 committed:
more comments

ashishkg0022 committed:
comments in parse

ashishkg0022 committed:
explained properly basic structure

ashishkg0022 committed:
updated __init__

ashishkg0022 committed:
updated init

Abdullahjavednesar commented:
@ashishkg0022 for those utility functions which you've recently modified have you added relevant tests for all (which probably was failing before)? Although I've seen your work before looks good to me, but if you are missing anywhere just recheck.

ashishkg0022 commented:
Yes, I think I have added test cases for them.

ashishkg0022 commented:
There are some more things that need to be done to improve utility functions. 
I have opened a separate PR especially for improving utility functions. ( #14956)

ashishkg0022 commented:
I apologise, I have not run pyflakes on it. I was not aware of it (pyflakes). I will run it today and fixed if something is undefined. I will fix it today. Then this can be merged.

ashishkg0022 committed:
fixed undefinef issues after running pyflakes

ashishkg0022 commented:
Fixed few undefined error using pyflakes.

certik commented:
@Upabjojr would you mind reviewing this PR please and give it a final +1?

Upabjojr commented:
@ashishkg0022 did you add the instructions on how to run the tests?

Upabjojr commented on sympy/integrals/rubi/__init__.py:
```diff
@@ -24,10 +69,10 @@
+Parsing Rules and Tests
+=======================
+Code for parsing rule and tests are included in sympy.
+They have been properly explained with steps in `rubi_parsing_guide.md` in parsetools.
```
full path of `rubi_parsing_guide.md` is better

Upabjojr commented on sympy/utilities/runtests.py:
```diff
@@ -671,6 +671,7 @@ def _doctest(*paths, **kwargs):
         "sympy/physics/unitsystems.py", # raises deprecation warning
         "sympy/parsing/latex/_antlr/latexlexer.py", # generated code
         "sympy/parsing/latex/_antlr/latexparser.py", # generated code
+        "sympy/integrals/rubi/rubi.py"
```
are we blacklisting the doctests as well?

ashishkg0022 replied:
only for `rubi.py`, not for utility_functions. 
Actually, the loading time is around 10 - 11 minutes. So, travis fails. If nothing happens in 10 minutes, travis fails. So I have blacklisted it.

ashishkg0022 committed:
added instructions for running tests

ashishkg0022 committed:
solved merge conflicts

Upabjojr commented:
I guess the master branch should be merged into this PR.

ashishkg0022 commented:
> I guess the master branch should be merged into this PR.

I have merged the master branch

ashishkg0022 commented:
The travis failed due to some network error.

Upabjojr commented:
Some of the rubi tests keep failing. For example, `test_1_3.py`, is it supposed to be like that?

ashishkg0022 commented:
`test_1_3.py` is from last year. I have not run it due to a large time taken by it. 
What is the status of newly added test files? Do a lot of them fail?

Upabjojr commented:
> Do a lot of them fail?

I didn't test them all so far.

ashishkg0022 commented:
Can you please run them all? I have tested them in part. I have not run them all together. Like, when I tested exponentials they passed. But after a week I made some changes in trigo rules and utility functions. So it might be possible very few might fail because of some changes done later. For any new change, it's not possible to run all tests again.

asmeurer commented:
Regarding the release notes, I accidentally added `rubi` as a submodule, but since it's been moved to `integrals.rubi`, the header should just be "integrals" (I've fixed it). c.f. https://github.com/sympy/sympy-bot/pull/16. If there end up being a lot of integrals release notes we can split them up into different subsubmodules like integrals.risch, integrals.rubi, and so on.

Upabjojr commented:
For the rubi_test results:
https://gist.github.com/Upabjojr/02cf3005ec995c7401418624a4bc11c7

ashishkg0022 commented:
I saw it. Tests which I have run recently like hyperbolic are passing but a few of trinomials and miscellaneous algebra might be failing. Is there some way I can get a list of all expressions in the test set which are failing?

Upabjojr commented:
You have the line numbers. Some tests are actually still running.
